lte/gateway/c/oai/tasks/sgw/sgw_handlers.c

```diff
@@ -685,145 +726,176 @@ int sgw_handle_sgi_endpoint_deleted(
+    for (idx = 0;
+         idx <
+         modify_bearer_pP->bearer_contexts_to_be_modified.num_bearer_context;
+         idx++) {
+      eps_bearer_ctxt_p = sgw_cm_get_eps_bearer_entry(
+          &bearer_ctxt_info_p->sgw_eps_bearer_context_information
+               .pdn_connection,
+          modify_bearer_pP->bearer_contexts_to_be_modified.bearer_contexts[idx]
               .eps_bearer_id);
-      message_p->ittiMsgHeader.imsi = imsi64;
-      rv = send_msg_to_task(&spgw_app_task_zmq_ctx, TASK_MME, message_p);
-      OAILOG_FUNC_RETURN(LOG_SPGW_APP, rv);
-    } else {
-      // Send end marker to eNB and then delete the tunnel if enb_ip is
-      // different
-      if (is_enb_ip_address_same(
-              &modify_bearer_pP->bearer_contexts_to_be_modified
-                   .bearer_contexts[0]
-                   .s1_eNB_fteid,
-              &eps_bearer_ctxt_p->enb_ip_address_S1u) == false) {
-        struct in_addr ue  = eps_bearer_ctxt_p->paa.ipv4_address;
-        struct in_addr enb = {.s_addr = 0};
+
+      if (NULL == eps_bearer_ctxt_p) {
+        sgi_update_end_point_resp.bearer_contexts_not_found[sgi_rsp_idx++] =
+            modify_bearer_pP->bearer_contexts_to_be_modified
+                .bearer_contexts[idx]
+                .eps_bearer_id;
+        sgi_update_end_point_resp.num_bearers_not_found++;
+      } else {  // eps_bearer_ctxt_p found
         enb.s_addr =
             eps_bearer_ctxt_p->enb_ip_address_S1u.address.ipv4_address.s_addr;
 
-        // This is best effort, ignore return code.
-        gtp_tunnel_ops->send_end_marker(enb, modify_bearer_pP->teid);
-
-        OAILOG_DEBUG_UE(
-            LOG_SPGW_APP, imsi64, "Delete GTPv1-U tunnel for sgw_teid : %d\n",
-            eps_bearer_ctxt_p->s_gw_teid_S1u_S12_S4_up);
-        // delete GTPv1-U tunnel
-        rv = gtp_tunnel_ops->del_tunnel(
-            ue, eps_bearer_ctxt_p->s_gw_teid_S1u_S12_S4_up,
-            eps_bearer_ctxt_p->enb_teid_S1u, NULL);
-        if (rv < 0) {
-          OAILOG_ERROR_UE(LOG_SPGW_APP, imsi64, "ERROR in deleting TUNNEL\n");
+        // Send end marker to eNB and then delete the tunnel if enb_ip is
+        // different
+        if (is_enb_ip_address_same(
+                &modify_bearer_pP->bearer_contexts_to_be_modified
+                     .bearer_contexts[idx]
+                     .s1_eNB_fteid,
+                &eps_bearer_ctxt_p->enb_ip_address_S1u) == false) {
+          struct in_addr ue = eps_bearer_ctxt_p->paa.ipv4_address;
+
+          OAILOG_DEBUG_UE(
+              LOG_SPGW_APP, imsi64, "Delete GTPv1-U tunnel for sgw_teid : %d\n",
+              eps_bearer_ctxt_p->s_gw_teid_S1u_S12_S4_up);
+          // delete GTPv1-U tunnel
+          rv = gtp_tunnel_ops->del_tunnel(
+              ue, eps_bearer_ctxt_p->s_gw_teid_S1u_S12_S4_up,
+              eps_bearer_ctxt_p->enb_teid_S1u, NULL);
+          // This is best effort, ignore return code.
+          gtp_tunnel_ops->send_end_marker(enb, modify_bearer_pP->teid);
```

Comment:
The end marker should be sent before the tunnel rule is deleted. Why did the ordering has changed here?